### PR TITLE
ATB-65 Define and create CloudWatch dashboard

### PR DESCRIPTION
**Pull request by MinaMahdi**
## What? & Testing 

Created a Dashboard titled AMFCloudWatchDashboard that includes metric graphs for ECS, ELB, and DynamoDB. 
Screenshot 2023-02-17 at 15 15 40

Each Metric Graph: 
- 'API Gateway 4xx, 5xx Errors and Total Invocations' illustrates an overview of the amount of 4xx and 5xx errors in comparison to the total number of API Gateway invocations. A surge in either 4xx and 5xx errors for a set amount of time is a cause for concern to be investigated. Experiencing more requests than usual could indicate an endpoint returning an error message causing multiple re-entries on client side or if there are also less requests than usual, there could be a permissions issues withholding API calls. 

Screenshot 2023-02-16 at 14 40 11

- 'API Gateway Latency' illustrates the maximum latency measured in milliseconds . This is important to track to ensure there is no significant lack of delay when the API received the request to when it responds impacting performance.
Screenshot 2023-02-16 at 14 39 26

- ' ECS Auto Scaling (Production Only)' illustrates the sum number of API Gateway invocations, and maximum ECS CPUUtilization and MemoryUtilization. 

Screenshot 2023-02-16 at 15 14 50

 - 'ECS Running Task Count, Desired and Pending Task Count' illustrates the number of tasks in the ECS Service that are currently running, number of tasks pending and the desired amount of tasks that have been scheduled to run as indicated in the task definition. Important to highlight that tasks are running as expected. For example if we see a drop in the ratio of Running Task Count to Desired Task Count could potentially indicate ECS is running into a resource constraint. If a Task is pending and has disappeared could indicated that a Task may have failed an ELB health check. 

Screenshot 2023-02-17 at 19 10 26

 - 'ECS Cluster Service Count' illustrates the number of services are running in a cluster. Potentially ECS Service will have this autoscaling enabled in Production; currently ECS Service only as one task attached to it that is running, but in the future presume there will be more, so having this dashboard would be useful to see how many services will be created with demand. 
 
Screenshot 2023-02-17 at 07 59 13

- 'Number of requests made to ELB vs Response Time' illustrates the amount of requests made to the load balancer and the amount of time that it takes for the host to return a response via load balancer. A potential drop in requests could indicate a problem from client end where they are attempting to gain contact with the load balancer, but it's returning errors and client is re-trying, hence a potential surge in number of requests.  

Screenshot 2023-02-17 at 19 12 01

- 'ALB endpoint functionality' monitors ActiveConnectionCount along with ProcessedBytes. ProcessedBytes will monitor the number of bytes processed by the Load Balancer and compare it to ActiveConnectionCount which displays the number of TCP connections at a given point in time ( connections between client and Load Balancer, along with Load Balancer and Target) , however this metric does not differentiate between connections from the client and to the target. 

Screenshot 2023-02-21 at 11 27 22

- 'ALB 502,503 and 504 errors' illustrates the number of HTTP 502's, 503, and 504 errors that stem from the Load Balancer.  This will provide a deeper understanding of the type of server error for troubleshooting purposes.  

Screenshot 2023-02-21 at 11 28 18

- 'ALB Target Health Check' compares HealthyHostCount with UnHealthyHostCount to showcase TargetGroup is running optimally and receiving requests. Potentially useful to create an alarm for Unhealthy hosts based on the metric. A failed target group health check will not route requests. If UnhealthyHostCount is high you need to investigate and maintain healthy instances in each availability zone for performance purposes. 

Screenshot 2023-02-21 at 11 29 24

- 'Unsuccessful Connections Between Load Balancer And Host' monitors the metric TargetConnectionErrorCount. A high count can cause latency for client since the Load Balancer will re-try the request to the target. Connection errors can happen due to instances being overloaded and rejecting new connections. 

Screenshot 2023-02-21 at 11 30 29

- 'DynamoDB Read and Write Capacity' is useful to showcase a pattern, rather than the actual data itself from one point of view as we are not able to create a dashboard to indicate how many sessions are stored in DynamoDB so this could be helpful to see if there is a huge spike that doesn't follow the pattern indicating some type of malicious activity or whether there is a spike during certain times of the day too. 

Screenshot 2023-02-17 at 19 18 36

- 'DynamoDB Average Latency' monitors the latency of PutItem, GetItem, UpdateItem, DeleteItem. Important to monitor if there is a higher latency at a specific point in the journey to resolve issues. 

Screenshot 2023-02-17 at 19 17 33

## Why? 

These dashboard will allow you to monitor resource behaviour for AMF to verify that is performing as expected.

*Note: WAF dashboard were not able to be created due to organisational error and so unable to create a cross-region dashboard as illustrated in the image below so will be creating WAF dashboards in Grafana or Splunk potentially if that suits.  
WAF logging admin error

### CODE REVIEW

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:362]: Could we move the `APIGatewayIDParameter` resource to the API Gateway section please

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:376]: And could we put this resource just beneath the ContainerService resource please

File: deploy/template.yaml
Comment (MinaMahdi) [reply deploy/template.yaml]: Thanks Dom all edited now 👍

**Comment by TaherKapasi** [comment]
Hi @MinaMahdi - please update  the What and the Why sections of this PR to reflect intention of the solution - i.e what does it achieve, what does it affect (if anything) etc and add all the things listed below

@domantas-gds - could we enhance the PR template that poses all the import questions and some tick boxes to ensure mandatory steps inc screenshots of deployment events, components and services configs, log outputs etc inc links to docs changes and instructions, jiras, READMe added/updated and so on

**Comment by TaherKapasi** [comment]
Hi @MinaMahdi 

Where are the dashboards for

- ECS cluster, service, task level dashboard
- ALB metrics
- Applications logs (metrics on log levels produced by the container, e.g how many ERROR, WARN log messages per period.

Could you also add a dashboard for db metrics for read/write usage vs limits, item counts (if they are available etc) to indicate the health of the session storage table and index .....   also add this req into this PR's jira ticket -- fyi @domantas-gds 

PS - could you squash all these commits (interactive rebase fixup) into one commit please

**Review by TaherKapasi — CHANGES_REQUESTED** [review]

File: deploy/template.yaml
Comment (domantas-gds) [reply deploy/template.yaml:1042]: No please reinstate the IsProduction condition

File: deploy/template.yaml
Comment (domantas-gds) [reply deploy/template.yaml:1438]: Yes required for references to the API Gateway, ECS Cluster and Container Service resource return values within the template

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:1450]: Could interpolate this to ${AWS::Region} and should work as we're providing the !Sub function for entire dashboard body

Comment (domantas-gds) [reply deploy/template.yaml:1450]: Please be sure to update all other region references thanks

File: deploy/template.yaml
Comment (MinaMahdi) [reply deploy/template.yaml]: I've reinstated 👍

File: deploy/template.yaml
Comment (MinaMahdi) [reply deploy/template.yaml]: Edited thanks 👍

**Review by TaherKapasi — APPROVED** [review]

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:1599]: nit: ALB as it's an application load balancer

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:1619]: ALB :)

File: deploy/template.yaml
Comment (domantas-gds) [inline deploy/template.yaml:1580]: nit: ALB for consistency?

File: deploy/template.yaml
Comment (MinaMahdi) [reply deploy/template.yaml]: Edited now thanks!

**Review by domantas-gds — DISMISSED** [review]

**Comment by huwd** [comment]
@MinaMahdi @domantas-gds and @TaherKapasi can we merge this?

**Comment by MinaMahdi** [comment]
> @MinaMahdi @domantas-gds and @TaherKapasi can we merge this?
Hi Huw, sorry for the late reply. If it's okay with you, just want @TaherKapasi to approve when he gets a chance before merging in. Thanks :)

**Review by domantas-gds — APPROVED** [review]

**Review by TaherKapasi — APPROVED** [review]

**Review by huwd — APPROVED** [review]

**Pull request merged**
